Unit 9: Model optimization (Introduction to Model Optimization for Deployment) (#117)
johko/computer-vision-course: chapters/en/Unit 9 - Model Optimization/intro_to_model_optimization.mdx

# Introduction to model optimization for deployment

> Have you ever felt confused after the model training stage? What else should you do? If yes, this chapter will help you. In general, the step after we have trained a computer vision model is to deploy it so that other people can use our model. However, when the model has successfully deployed in production, many problems arise, such as the model size is too large, the prediction process taking a long time, and limited memory on the device. These problems can happen because we usually deploy models on devices with smaller specifications than the hardware for training. To overcome the issues, we can carry out additional stages before deploying, model optimization.

Reviewer: ```suggestion
Have you ever felt confused after the model training stage? What else should you do? If yes, this chapter will help you. In general, the step after we have trained a computer vision model is to deploy it so that other people can use our model. However, when the model has successfully deployed in production, many problems arise, such as the model size being too large, the prediction process taking a long time, and limited memory on the device. These problems can happen because we usually deploy models on devices with smaller specifications than the hardware for training. To overcome the issues, we can carry out additional stages before deploying and model optimization.
```

## What is model optimization?

> Model optimization is a process of modifying an AI model that we made to make it better in terms of performance. These modifications are crucial because the hardware we use during training and inference will be very different in most cases. The hardware specifications at the time of inference are smaller, which is why this optimization model needs to be carried out. For example, we have training on high-performance GPUs, and the model inference process will run on edge devices (e.g., microcomputers, mobile devices, IoT, etc.). Of course, these devices have different specifications and tend to be smaller. Carrying out model optimization is crucial so our model can run smoothly on devices with lower specifications.

Reviewer: ```suggestion
Model optimization is a process of modifying a model we trained to make it better in terms of efficiency. These modifications are crucial because the hardware we use during training and inference will be very different in most cases. The hardware specifications at the time of inference are smaller, which is why this optimization model needs to be carried out. For example, we have training on high-performance GPUs, and the model inference process will run on edge devices (e.g., microcomputers, mobile devices, IoT, etc.). Of course, these devices have different specifications and tend to be smaller. Carrying out model optimization is crucial so our model can run smoothly on devices with lower specifications.
```
When we talk about performance we often mention about accuracy and so on so it's best if we say efficiency

## Why is it important for deployment in computer vision?

> As we already know, the optimization model is important in the pre-deployment stage, but why? Several things make this optimization model important to do before the deployment stage. Some of these things are:

Reviewer: ```suggestion
As we already know, optimizing the model is important in before the deployment stage, but why? Several things make this optimization model important to do before the deployment stage. Some of these things are:
```

2. Latency requirements: Many computer vision applications, such as self-driving cars and augmented reality, require real-time response. This means the model must be able to process data and generate results quickly. Optimization can significantly increase the inference speed of a model and ensure it can meet latency constraints.
3. Power consumption: Devices that use batteries, such as drones and wearable devices, require models with efficient power usage. Optimization techniques can also reduce battery consumption which is often caused by model sizes that are too large.
4. Hardware compatibility: Sometimes, different hardware has its capabilities and limitations. Several optimization techniques are specifically used for specific hardware. If this is done, we can easily overcome the hardware limitations.

## Different types of model optimization techniques

> There are several techniques in the model optimization, which will be explained later in the next section. However, this section will briefly describe several types:

Reviewer: ```suggestion
There are several techniques in the model optimization, which will be explained in the next section. However, this section will briefly describe several types:
```

1. Pruning: Pruning is the process of eliminating redundant or unimportant connections in the model. This aims to reduce model size and complexity.
![Pruning](https://huggingface.co/datasets/hf-vision/course-assets/raw/main/pruning.png)
2. Quantization: Quantization means converting model weights from high-precision formats (e.g., 32-bit floating-point) to lower-precision formats (e.g., 16-bit floating-point or 8-bit integers) to reduce memory footprint and increase inference speed.
3. Knowledge Distillation: Knowledge distillation aims to transfer knowledge from a complex and larger model (teacher model) to a smaller model (student model) by mimicking the behavior of the teacher model.
![Knowledge Distillation](https://huggingface.co/datasets/hf-vision/course-assets/raw/main/knowledge_distillation.png)
4. Low-rank approximation: Approximates large matrices with small ones, reducing memory consumption and computational costs.

> 5. Model compression with hardware accelerators: This process is like pruning and quantization, specifically on specific hardware such as NVIDIA GPUs and Intel Hardware.

Reviewer 2: "specifically on specific" is a bit repetitive.

Author: Just change it

## Trade-offs between accuracy, performance, and resource usage
A trade-off exists between accuracy, performance, and resource usage when deploying a model. That's when we have to decide which part to prioritize so that the model can be maximized in the case at hand.

> 1. Accuracy is the model's ability to predict correctly. High accuracy is needed in all applications, which also causes higher performance and resource usage. Complex models with high accuracy usually require a lot of memory, so there will be limitations if they are deployed on resource-constrained devices. Complex models with high accuracy usually require a lot of memory, so there will be limitations if they are deployed on resource-constrained devices.

Reviewer 2: Accuracy seems to be used here as a broad term for model performance in regards to evaluation metrics, however, it is also a very specific classification metric. This may be confusing. In some cases, the "accuracy" metric is not necessarily what we care about.

Author: Yep, I also realize it when I write. But, I still searching another term like accuracy which means performance metrics of the model 🤔

Reviewer 2: Maybe something like "predictions quality" defined by the metrics used in each particular case?

> 2. Performance is a model of speed and efficiency (latency). This is important so the model can make predictions quickly, even in real time. However, optimizing performance will usually result in decreasing accuracy.

Reviewer 2: ```suggestion
2. Performance is the model's speed and efficiency (latency). This is important so the model can make predictions quickly, even in real time. However, optimizing performance will usually result in decreasing accuracy.
```

3. Resource usage is the computational resources needed to perform inference on the model, such as CPU, memory, and storage. Efficient resource usage is crucial if we want to deploy models on devices with certain limitations, such as smartphones or IoT devices.

Image below shows a common computer vision model in terms of model size, accuracy, and latency. Bigger model has high accuracy, but needs more time for inference and big size.

![Model Size VS Accuracy](https://huggingface.co/datasets/hf-vision/course-assets/raw/main/model_size_vs_accuracy.png)

![Accuracy VS Latency](https://huggingface.co/datasets/hf-vision/course-assets/raw/main/accuracy_vs_latency.png)

> These are the three things we must consider: where will the model we create focus be? For example, focusing on high accuracy will result in a slower model during inference or require extensive resources. To overcome this, we apply one of the optimization methods as explained so that the model we get can maximize or balance the three trade-offs above (accuracy, performance, resource usage)

Reviewer: ```suggestion
These are the three things we must consider: where do we focus on the model we trained? For example, focusing on high accuracy will result in a slower model during inference or require extensive resources. To overcome this, we apply one of the optimization methods as explained so that the model we get can maximize or balance the trade-off between the three components mentioned above.
```
